system-parachains/coretime/coretime-polkadot/src/migrations.rs

```diff
@@ -0,0 +1,352 @@
+// Copyright (C) Parity Technologies and the various Polkadot contributors, see Contributions.md
+// for a list of specific contributors.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Migration to fix the coretime migration lease offset-related issues
+//!
+//! This fixes a problem with the leases where the relay migration did not take the lease offset
+//! into consideration, so the end of leases is 64 days short, in some cases leading to them being
+//! dropped completely.
+
+use crate::{weights, Runtime, RuntimeOrigin};
+use frame_support::{pallet_prelude::*, traits::OnRuntimeUpgrade};
+use pallet_broker::{
+	CoreAssignment, CoreAssignment::Pool, CoreMask, LeaseRecordItem, Leases, LeasesRecordOf,
+	PotentialRenewalId, PotentialRenewals, Schedule, ScheduleItem, WeightInfo, Workplan,
+};
+
+use sp_std::vec::Vec;
+
+#[cfg(feature = "try-runtime")]
+use pallet_broker::{CoreAssignment::Task, PotentialRenewalRecord, SaleInfo, SaleInfoRecordOf};
+#[cfg(feature = "try-runtime")]
+use sp_runtime::TryRuntimeError;
+
+/// The log target.
+const TARGET: &str = "runtime::bootstrapping::fix-migration";
+
+// Alias into the broker weights for this runtime.
+type BrokerWeights = weights::pallet_broker::WeightInfo<Runtime>;
+
+pub struct FixMigration;
+
+impl OnRuntimeUpgrade for FixMigration {
+	fn on_runtime_upgrade() -> Weight {
+		if PotentialRenewals::<Runtime>::get(PotentialRenewalId { core: 6, when: 292605 }).is_none()
+		{
+			// Idempotency check - this core will never be renewable at this timeslice ever again.
+			log::error!(target: TARGET, "This migration includes hardcoded values not relevant to this runtime. Bailing.");
+			return <Runtime as frame_system::Config>::DbWeight::get().reads(1);
+		}
+
+		// Add leases for 2040, 2094, 3333, 2106, 2101, 2093 with a properly calculated end
+		// timeslice. Add 11520 for all other leases.
+		let leases: LeasesRecordOf<Runtime> = LEASES
+			.iter()
+			.map(|(until, task)| LeaseRecordItem { until: *until, task: *task })
+			.collect::<Vec<_>>()
+			.try_into()
+			.expect("Within range of bounded vec");
+		Leases::<Runtime>::put(leases);
+
+		// This reorders the cores, so the existing entries to the workplan need to be overwritten.
+		for (&(para_id, _), core_id) in LEASES.iter().zip(51u16..56u16) {
```

Comment:
I think there have been some regressions since I left this PR, once I finish going through these comments I'll review in full, but I expect that the regressions have been fixed now